src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CompositeValueProvider.cs

```diff
@@ -184,5 +205,63 @@ public IValueProvider Filter(BindingSource bindingSource)
 
             return new CompositeValueProvider(filteredValueProviders);
         }
+
+        /// <inheritdoc />
+        /// <remarks>
+        /// Value providers are included by default. If a contained <see cref="IValueProvider"/> does not implement
+        /// <see cref="IKeyRewriterValueProvider"/>, <see cref="Filter()"/> will not remove it.
+        /// </remarks>
+        public IValueProvider Filter()
+        {
+            var shouldFilter = false;
+            for (var i = 0; i < Count; i++)
+            {
+                var valueProvider = Items[i];
+                if (valueProvider is IKeyRewriterValueProvider)
+                {
+                    shouldFilter = true;
+                    break;
+                }
+            }
+
+            if (!shouldFilter)
+            {
+                // No inner IKeyRewriterValueProvider implementations. Nothing to exclude.
+                return this;
+            }
+
+            var filteredValueProviders = new List<IValueProvider>();
+            for (var i = 0; i < Count; i++)
+            {
+                var valueProvider = Items[i];
+                if (valueProvider is IKeyRewriterValueProvider keyRewriterValueProvider)
+                {
+                    var result = keyRewriterValueProvider.Filter();
+                    if (result != null)
+                    {
+                        filteredValueProviders.Add(result);
+                    }
+                }
+                else
+                {
+                    // Assume value providers that aren't rewriter-aware do not rewrite their keys.
+                    filteredValueProviders.Add(valueProvider);
+                }
+            }
+
+            if (filteredValueProviders.Count == 0)
+            {
+                // Do not create an empty CompositeValueProvider.
+                return null;
+            }
+
+            if (filteredValueProviders.Count == Count)
```

Comment:
This doesn't look correct. `keyRewriterValueProvider.Filter` may return a different value provider (consider nested `CompositeValueProvider`) so you could get the same count but have different values.

Reply:
This case is identical to the unchanged part of `Filter(BindingSource)` (for `IBindingSourceValueProvider`) [above](https://github.com/aspnet/Mvc/pull/7475/files#diff-6ec1ce373e87fedc0bfa03f033f1e0c4R200). The nested `CompositeValueProvider` and the inner-returns-different-`IValueProvider` cases will hit bugs.

That said, it's a very subtle part of these methods' behaviours that they return themselves in some cases. Our code certainly doesn't take advantage of this. This and the similar `Filter(BindingSource)` block might be a premature optimization we could remove without causing inconsistencies between `CompositeValueProvider` and other providers and without even a smidgen of back-compatibility issues.

@rynowak I'm leaning toward taking the unchanged-`Count` stuff out both for generality (the different value provider case @pranavkm mentioned) and because `CompositeValueProvider` isn't much bigger than the `List` it's constructed with. **Thoughts?**